### PR TITLE
Added tutorial to optimize the DOS energy reference during training

### PR DESCRIPTION
**Pull request by HowWeiBin**
Included a tutorial that shows how to optimize the DOS energy reference during training using Rascaline SOAP power spectrum and pytorch. 

The performance is compared against 2 other models, trained on standard energy references

### CODE REVIEW

File: examples/dos-align/dos-align.py [hunk 1]
```diff
@@ -0,0 +1,910 @@
...
+# %%
+#
+# Step 1: Constructing the DOS with different energy references
+# ------------------------------------------------------------------------------------------
+# 1) Construct the DOS using the original reference
+# 2) Calculate the Fermi level from the DOS
+# 3) Build a set of eigenenergies, with the energy reference set to the fermi level
+# 4) Truncate the DOS energy window so that the DOS is well-defined at each point
+# 5) Construct the DOS in the truncated energy window under both references
+# 6) Construct Splines for the DOS to facilitate interpolation during model training
```
Comment (jwa7) [inline examples/dos-align/dos-align.py:100]: It would be useful to use these below for each subsection

File: examples/dos-align/environment.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+channels:
+  - conda-forge
+dependencies:
+  - python=3.11
+  - pip
+  - rust > 1.65
+  - pip:
+    - ase
+    - matplotlib
+    - metatensor
```
Comment (jwa7) [inline examples/dos-align/environment.yml]: Not sure if this is needed

File: examples/dos-align/environment.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    - rascaline @ git+https://github.com/Luthaf/rascaline@ca957642f512e141c7570e987aadc05c7ac71983
```
Comment (jwa7) [inline examples/dos-align/environment.yml:10]: Could use a newer version perhaps?

File: examples/dos-align/dos-align.py [hunk 4]
```diff
@@ -0,0 +1,910 @@
...
+# gaussian broadening, the energy axis of the DOS extends
+# 3eV wider than the range of values for the eigenenergies
+energy_lower_bound = minE - 1.5
+energy_upper_bound = maxE + 1.5
+
+# Gaussian Smearing for the eDOS, 0.3eV is the appropriate value for this dataset
+
+sigma = torch.tensor(0.3)
+
+# energy axis, with a grid interval of 0.05 eV
```
Comment (jwa7) [inline examples/dos-align/dos-align.py:121]: Might be helpful to have this is a variable

File: examples/dos-align/dos-align.py [hunk 5]
```diff
@@ -0,0 +1,910 @@
...
+    total_edos.append(e_dos)
+
+total_edos = torch.vstack(total_edos)
+total_edos = (total_edos.T * normalization).T
+
+print(f"The final shape of all the DOS in the dataset is: {list(total_edos.shape)}")
+
+# %%
+# Now we calculate the Fermi Level by integrating the DOS
+# and then use cubic interpolation and brentq to determine the fermi level
```
Comment (jwa7) [inline examples/dos-align/dos-align.py]: An explanation of the fact that our truncation range does not include deep states might be nice here, such that we are integrating up to 4 (=/= n_electrons)

File: examples/dos-align/dos-align.py [hunk 6]
```diff
@@ -0,0 +1,910 @@
...
+print(f"The lowest eigenenergy using the fermi level energy reference is {minE_Ef:.3}")
+print(f"The highest eigenenergy using the fermi level energy reference is {maxE_Ef:.3}")
+
+
+# %%
+# With the fermi levels, we can also truncate the energy window for DOS prediction.
+# In this example, we truncate the energy window such that it is 3eV above
+# the highest Fermi level in the dataset
+
+# For the Average Hartree Potential energy reference
```
Comment (jwa7) [inline examples/dos-align/dos-align.py:211]: Maybe emphasise that we're generating different datasets: i.e first we align to the Hartree, and we also align to fermi

File: examples/dos-align/dos-align.py [hunk 7]
```diff
@@ -0,0 +1,910 @@
...
+
+    return value: shape of (n x x)
+
+    x : shape of (n x n_points)
+    """
+    interval = torch.round(
+        spline_positions[1] - spline_positions[0], decimals=4
+    )  # get spline grid intervals
+    x = torch.clamp(
+        x, min=spline_positions[0], max=spline_positions[-1] - 0.0005
```
Comment (jwa7) [inline examples/dos-align/dos-align.py:328]: A comment on why 0.0005 is subtracted here?

File: examples/dos-align/dos-align.py [hunk 8]
```diff
@@ -0,0 +1,910 @@
...
+# %%
+#
+# We feed the Hyperparameters into rascaline to compute the SOAP Power spectrum
+
+
+calculator = SoapPowerSpectrum(**HYPER_PARAMETERS)
+R_total_soap = calculator.compute(structures)
+R_total_soap.keys_to_samples("species_center")
+R_total_soap.keys_to_properties(["species_neighbor_1", "species_neighbor_2"])
+
```
Comment (jwa7) [inline examples/dos-align/dos-align.py:411]:
```suggestion
# Extract the data tensor from the single block
```

File: examples/dos-align/dos-align.py [hunk 9]
```diff
@@ -0,0 +1,910 @@
...
+                idx,
+                self.n_atoms_per_structure[idx],
+            )
+
+
+def collate(
+    batch,
+):  # Defines how to collate the outputs of the __getitem__ function at each batch
+    for x, idx, index in batch:
+        return (x, idx, index)
```
Comment (jwa7) [inline examples/dos-align/dos-align.py:510]:
```suggestion
    return tuple(batch)
```

File: examples/dos-align/dos-align.py [hunk 10]
```diff
@@ -0,0 +1,910 @@
...
+        self.align = align
+        if align:
+            initial_alignment = torch.zeros(n_train)
+            self.alignment = torch.nn.parameter.Parameter(initial_alignment)
+
+    def forward(self, x):
+        result = self.fc1(x)
+        result = self.silu(result)
+        result = self.fc2(result)
+        result = torch.exp(result)
```
Comment (jwa7) [inline examples/dos-align/dos-align.py]: A la discussion
```suggestion
```

**Review by jwa7 — CHANGES_REQUESTED** [review]
Looking good! Few comments below. Also, does this example definitely get built with nox? I don't see `Documentation / generate-example (dos-align)` below

File: examples/dos-align/dos-align.py (hunk 9 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py:510]: Somehow it still breaks when i use the tuple(batch) method, will revert back to original

File: examples/dos-align/dos-align.py (hunk 1 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py:100]: Done!

File: examples/dos-align/environment.yml (hunk 2 above)
Comment (HowWeiBin) [reply examples/dos-align/environment.yml]: Removed metatensor

File: examples/dos-align/environment.yml (hunk 3 above)
Comment (HowWeiBin) [reply examples/dos-align/environment.yml:10]: Stuck to this version for now

File: examples/dos-align/dos-align.py (hunk 4 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py:121]: Done

File: examples/dos-align/dos-align.py (hunk 5 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py]: Done

File: examples/dos-align/dos-align.py (hunk 6 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py:211]: Done

File: examples/dos-align/dos-align.py (hunk 7 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py:328]: Done

File: examples/dos-align/dos-align.py (hunk 8 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py:411]: Done

File: examples/dos-align/dos-align.py (hunk 10 above)
Comment (HowWeiBin) [reply examples/dos-align/dos-align.py]: Done

**Review by jwa7 — APPROVED** [review]
Looks great, thank you!

**Pull request merged**